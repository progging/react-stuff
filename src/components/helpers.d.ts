export declare const spacingCss: (cssAttribute: string, baseUnitPx: number, spacing?: number | number[] | undefined) => string;
