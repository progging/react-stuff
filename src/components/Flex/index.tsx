import React from 'react'
import styled from 'styled-components'
import kebabCase from 'lodash.kebabcase'

const Flex: any = styled.div`
  display: ${(props: any) => props.inlineFlex ? 'inline-flex' : 'flex'};
  ${(props: any) => props.css.map((prop: any) => `${prop.cssProp}: ${prop.val};`).join(' ')}
`

type GlobalCssProps = 'inherit' | 'initial' | 'unset'

export type FlexboxCssProps = {
    alignItems?: 'strech' | 'center' | 'start' | 'end' | GlobalCssProps
    flexBasis?: any | GlobalCssProps
    flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse' | GlobalCssProps
    flexFlow?: any | GlobalCssProps
    flexGrow?: number | GlobalCssProps
    flexShrink?: number | GlobalCssProps
    flexWrap?: 'nowrap' | 'wrap' | 'wrap-reverse'
    inlineFlex?: boolean
    justifyContent?: 'start' | 'center' | 'space-between' | 'space-around' | 'space-evenly' | GlobalCssProps
    order?: number | GlobalCssProps
}

/***
 * Applies Flexbox styling.
 * @param props Flexbox properties should be passed as camelcase props on the component.
 */
export default function (props: FlexboxCssProps & any) {

    const {children, className, style, inlineFlex, ...cssProps} = props

    const css = Object.keys(cssProps).map(prop => ({cssProp: kebabCase(prop), val: cssProps[prop]}))

    return <Flex
        style={style}
        className={className}
        inlineFlex={inlineFlex}
        css={css}>
        {children}
    </Flex>
}