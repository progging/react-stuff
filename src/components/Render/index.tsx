import React from 'react'

type RenderProps = {
    children: any
    if: any
}

/***
 * Conditionally renders children if `condition` is truthy.
 * @param props
 */
export const Render = (props: RenderProps) => props.if && props.children