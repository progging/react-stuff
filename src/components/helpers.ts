export const spacingCss = (cssAttribute: string, baseUnitPx: number, spacing?: number | number[]) => {
    if (spacing === undefined) {
        return ''
    } else if (Array.isArray(spacing)) {
        if (spacing.length === 1 || spacing.length === 2 || spacing.length === 4) {
            return `${cssAttribute}: ${spacing.map((val: number) => `${(val * baseUnitPx)}px`).join(' ')};`
        }
        throw new Error('Array must contain either 1, 2, or 4 elements')
    } else {
        return `${cssAttribute}: ${spacing * baseUnitPx}px;`
    }
}