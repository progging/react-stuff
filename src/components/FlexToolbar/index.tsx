import React from 'react'
import Flex from '../Flex'

type ToolbarProps = {
    children: any
}

/***
 * Shorthand for `justify-content: space-between;` and `align-items: center;`
 * @param children
 */
export default function ({children}: ToolbarProps) {
    return (
        <Flex
            justifyContent="space-between"
            alignItems="center"
            style={{width: '100%'}}>
            {children}
        </Flex>
    )
}