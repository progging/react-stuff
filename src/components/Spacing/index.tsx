import React from 'react'
import styled from 'styled-components'
import {spacingCss} from '../helpers'

export type SpacingProps = {
    children: any
    padding?: number[]
    margin?: number[]
    theme?: any
}

const StyledSpacing = styled.div`
  ${(props: SpacingProps) => spacingCss('padding', props.theme ? props.theme.baseUnitPx : 8, props.padding)}
  ${(props: SpacingProps) => spacingCss('margin', props.theme ? props.theme.baseUnitPx : 8, props.margin)}
`

/***
 * Applies padding/margin in the magnitude of units (see docs on how to override what one unit is).
 * @param children
 * @param padding
 * @param margin
 */
export const Spacing = ({children, padding, margin}: any) => {
    return (
        <StyledSpacing
            padding={padding}
            margin={margin}>
            {children}
        </StyledSpacing>
    )
}