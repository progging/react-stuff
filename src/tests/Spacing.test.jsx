import React from "react"
import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import 'jest-styled-components'
import {ThemeProvider} from 'styled-components'
import {Spacing} from ".."
import {spacingCss} from '../components/helpers'

test("Component should have padding around it", () => {

    const component = mount(
        <Spacing padding={[1]}>
            <div/>
        </Spacing>
    )

    expect(mountToJson(component)).toMatchSnapshot()
    expect(component).toHaveStyleRule('padding', expect.any(String))
})

test("Component should adapt its spacing based on the theme", () => {

    const component = mount(
        <ThemeProvider theme={{baseUnitPx: 16}}>
            <Spacing padding={[1]}>
                <div/>
            </Spacing>
        </ThemeProvider>
    )

    expect(component).toHaveStyleRule('padding', '16px')
})

test("Spacing function must fail if spacing array does not have 1, 2, or 4 elements", () => {
    expect(() => {
        spacingCss('padding', 8, [])
    }).toThrowError(new Error('Array must contain either 1, 2, or 4 elements'))
    expect(() => {
        spacingCss('padding', 8, [1, 1, 1])
    }).toThrowError(new Error('Array must contain either 1, 2, or 4 elements'))
})