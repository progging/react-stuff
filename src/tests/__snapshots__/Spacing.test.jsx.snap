// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component should have padding around it 1`] = `
.c0 {
  padding: 8px;
}

<Spacing
  padding={
    Array [
      1,
    ]
  }
>
  <styled.div
    padding={
      Array [
        1,
      ]
    }
  >
    <div
      className="c0"
    >
      <div />
    </div>
  </styled.div>
</Spacing>
`;
